notebooks/Cross Validation Primer.ipynb

```diff
@@ -0,0 +1,432 @@
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "# There's No Such Thing as a Free Lunch\n",
+    "\n",
+    "In the field of machine learning, the \"no free lunch\" theorem \"state[s] that any two optimization \n",
+    "algorithms are equivalent when their performance is averaged across all possible problems\" (see [1] \n",
+    "and [2] for some references).That is, we can't tell which machine learning algorithm will best solve the problem in front of us. Of course, with more experience we'll be able to \"intuit\" which ones are better \n",
+    "suited than others, but a priori we can't say which is the best one.\n",
+    "\n",
+    "So we then need a way to measure if machine learning algorithm (A) is better than machine learning \n",
+    "algorithm (B). Or for that matter, if (A) with parameters {l,m,n} or (A) with parameters {p,q,r} will do better. This \n",
+    "is where **cross-validation** comes in.\n",
+    "\n",
+    "[1] Wolpert, David (1996), \"The Lack of A Priori Distinctions between Learning Algorithms\", Neural Computation, pp. 1341-1390\n",
+    "\n",
+    "[2] Wolpert, D.H., and Macready, W.G. (2005) \"Coevolutionary free lunches\", IEEE Transactions on Evolutionary Computation, 9(6): 721-735"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Cross-Validation\n",
+    "\n",
+    "Cross-validation is a way to measure how well a machine learning algorithm will generalize to more data (i.e the \"real\" dataset). For prediction problems, the steps typically go like this:\n",
+    "1. Prepare a set of data where the **target** (the value we want to predict) is already known. \n",
+    "2. Split this set into a **training set** and a **validation (or test) set**\n",
+    "3. Fit our model to the **training set**, and use the fit to predict values for the **validation set**\n",
+    "4. See how well the model predicted the **validation set**\n",
+    "\n",
+    "A typical measure is the **coefficient of determination, R^2**, which is defined as below:\n",
+    "\n",
+    "![R^2 definition](img/R2def.png)\n",
+    "\n",
+    "\n",
+    "The examples below go trough picking a good training set, demonstrating the need for cross validation, and showing\n",
+    "a techniquie called k-fold cross validation that is especially important when there is not a lot of data.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Picking a Training Set"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import sklearn as skl\n",
+    "import numpy as np\n",
+    "\n",
+    "import scripts.load_data as load\n",
+    "\n",
+    "import seaborn as sns\n",
+    "\n",
+    "%matplotlib inline"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
This cell is a bit long -- I try to avoid having more than a half a page of code in a row.

It'd be helpful to break this down and explain each step with a brief block of text, instead of using comments. It's a bit less intimidating and easier to follow.

For example, you explain the purpose of this chart only after it is generated -- a hint of what's to come should come before the cell that generates the chart. You might say something like "Cross-validation involves splitting the dataset into pieces, fitting on one piece, and testing performance on the other. How we make those splits is critically important for fitting a good model, as the cell below demonstrates."